### ISSUE TITLE
Where to put attribute for non-preemptable jobs

### ISSUE THREAD

**Issue by DaoWen** [twosigma/Cook#472]
I'm currently working on a feature that will allow users to specify that they don't want a job preempted, even if that means delaying the job until it can be run within the guaranteed resources in their share/quota.

Currently, I'm just using a boolean flag (`:job/guaranteed`) for development, but I don't think this is the way we'll want to record this attribute long-term. @dposada pointed out #291, which contains a lengthy discussion on a similar issue: "restartable" or "preemption-aversion" attributes for a job. However, that discussion didn't reach a solid conclusion since the issue was basically dropped.

I'd prefer having the "non-preemptable" option be something more extensible than just tacking another optional boolean flag onto the Job entities—but I'm not seeing any particularly attractive options right now.

Does anyone have suggestions on a better way to encode this option?

**Comment by dposada** [comment]
@wyegelwel Would this fit into the existing constraints field?

**Comment by shamsimam** [comment]
I like having the words preemptable and non-preemptable being used somehow.  
And also putting it in constraints.

If constraints do not work, maybe a field named `job/scheduling-mode` with the enum `preemptable` (default) and `non-premeptable`. 
We can then expand to contain other options later.

**Comment by wyegelwel** [comment]
Can you describe your current thinking on how to implement this?

In the past, we discussed this as deeper queue controls beyond (or in
addition to) priority. Where a user is able to dedicate chunks of their
share to different queues. Then users are allowed to define shares and
quotas on the subqueues to control how jobs in the queues are scheduled. If
a user had two sub queues, a and b, where a is given almost all the share
of the user and a quota that equals the share then jobs in a will be
scheduled before jobs in b, but won't be scheduled beyond the share of the
user. This provides to be non preemptible semantics.

However, the benefit of this approach is that it is more flexible then just
a non-preemptible option. A user might have an interactive queue and a
batch queue that get equal weight so that their interactive work won't
fully preempt their batch work.

Does that make sense? I'm typing on my phone so it is a little difficult to
go into detail.

**Comment by wyegelwel** [comment]
Let's collect use cases for this feature and related features.

**Comment by DaoWen** [comment]
I feel like these features might be orthogonal.

Having the ability to create and manage sub-queues allows for much more fine-grained control (as with your interactive/batch queues example); however, having sub-queues seems like it would add a *lot* of complexity to the Cook scheduler for a feature set that none of our users have actually asked for. Implementing the sub-queues would imply a lot of sub-tasks, which I think could easily take us multiple months to design, implement and test properly:

- Internal bookkeeping for tracking user queues in a user-friendly way. (Would we want to allow naming the sub-queues, or just give them UUIDs, or ...?)
- API for managing (creating/updating/removing) the sub-queues.
- API for querying the state of specific queues.
- Additional complexity in the scheduler for handling a single user's set of multiple queues separately. Maintaining global fairness in the presence of this feature might automatically follow from maintaining fairness among the a user's sub-queues—but that's not immediately clear to me.

I do like how the sub-queues idea introduces a kind of hierarchical homogeneity; i.e., the management of multiple sub-queues at the level of a single user basically looks the same as managing multiple users' queues at the level of the cluster. What that symmetry would give us a really clean conceptual model, I feel like a significant portion of the scheduler would need to be refactored in order to actually reuse code at multiple levels in the queue hierarchy.

Getting back to why the sub-queues and preemption-aversion might be orthogonal, I don't think dividing up resources among sub-queues automatically allows you to ensure a job is scheduled within your guaranteed share. If I understand the current implementation correctly, we have two different quota concepts in Cook: there's the guaranteed share (minimal resources each user should have access to when they actually go to run stuff), and there's the current dynamic share that's ≥ the guaranteed share based on current demand in the system. If we want to support the preemption-aversion use case via sub-queues, we need to allow users to allocate portions of the "guaranteed" and "transient" resources separately. E.g., for your A–B scenario, I'd might need to allocate 90% of my "guaranteed" and 0% of my "transient" resources to queue A, and then 10% "guaranteed" and 100% "transient" to queue B; otherwise, if any portion of A is considered "transient", then those job from A are game for preemption.

That's a lot of configuration to support preemption-aversion within sub-queues. I feel like it's much more natural to consider the two issues separately. If I have two queues, and I want a 9:1 resource split between them, I don't really want to concern myself with "guaranteed" vs "transient" resources—Cook should figure out what's fair and handle that for me; and if I want to put a preemption-averse job in one of my queues, I don't want to worry whether a given queue is composed of 100% "guaranteed" resources—I just want to tell Cook my job is probably going to die ungraciously if Cook kills it, and Cook should handle that as best it can.

**TL;DR:** While I think sub-queues *could* address the preemption-aversion use case, I personally feel that we should keep these concepts separate.

**Comment by DaoWen** [comment]
> Let's collect use cases for this feature and related features.

I should have included the details of the motivating use case in the first post. Sorry about that.

We have a user with a job that will leave the system in a bad state (unreleased mutex) if it's killed at an inopportune time. The user would like a way to mark the job as non-preemptable, implying that Cook won't schedule the job until it can give a reasonable guarantee that the job won't need to be preempted (killed) once it starts. Basically, it's a constraint that the job needs to fit within the "guaranteed" share of the user's resources, along with any other "guaranteed" jobs.

The user acknowledged that ideally he'd write his job to be more robust and avoid the mutex issue—but at the same time, this seems like a reasonable feature request.

______

We actually discussed an alternative solution to the user's problem as well. When preempting a job in the Cook executor, if we send a SIGINT first, followed by some delay, and then send the SIGKILL, that might make it a lot easier for users to shut down a preempted app gracefully. Trying to handle a SIGKILL that can arrive at any time might not even be possible to do cleanly, but the SIGINT would give some warning, and would probably be sufficient in most cases.

**Comment by DaoWen** [comment]
This feature is no longer in our roadmap. Closing the discussion for now.
